services/libs/integrations/src/integrations/discord/processStream.ts

```diff
@@ -0,0 +1,326 @@
+
+  const channel = await getMessages(
+    {
+      channelId: data.channelId,
+      token: getDiscordToken(ctx),
+      page: data.page,
+      perPage: 100,
+    },
+    ctx,
+  )
+
+  // if we are not in onboarding mode we need to check if we are not going too far in the past
+  if (!ctx.onboarding) {
+    // checking if we are not going too far in the past
+    const lastTimestamp = await getLastTimestampForChannel(ctx.stream.identifier, ctx)
+    if (lastTimestamp) {
+      const lastTimestampDate = new Date(lastTimestamp)
+      const lastTimestampSeconds = lastTimestampDate.getTime() / 1000
+      const nowSeconds = Date.now() / 1000
+      const diff = nowSeconds - lastTimestampSeconds
+      if (diff > MAX_RETROSPECT_SECONDS) {
+        ctx.log.warn(
+          `Going too far in the past for channel ${
+            ctx.stream.identifier
+          }. Last timestamp is ${lastTimestampDate.toISOString()}`,
+        )
+        // we don't need to parse next page, still need to parse threads
+        channel.nextPage = ''
+      }
+    }
+  }
+
+  if (channel.nextPage) {
+    await ctx.publishStream<DiscordChannelStreamData>(
+      `${DiscordStreamType.CHANNEL}:${data.channelId}:${channel.nextPage}`,
+      {
+        channelId: data.channelId,
+        guildId: data.guildId,
+        page: channel.nextPage,
+        channels: data.channels,
+      },
+    )
+  }
+
+  for (const record of channel.records as DiscordApiMessage[]) {
+    let parent: string | undefined
+    let parentChannel: string | undefined
+
+    let firstThreadMessage = false
+    // is the message starting a thread? if so we should get thread messages as well
+    if (record.thread) {
+      await ctx.publishStream<DiscordChannelStreamData>(
+        `${DiscordStreamType.CHANNEL}:${record.thread.id}`,
+        {
+          channelId: record.thread.id,
+          guildId: data.guildId,
+          page: '',
+          channels: data.channels,
+        },
+      )
+
+      firstThreadMessage = true
+      await cacheDiscordChannels(record.thread as DiscordApiChannel, ctx)
+    }
+
+    if (record.id === record.channel_id) {
+      firstThreadMessage = true
+    }
+
+    const channel = await getChannel(record.channel_id, getDiscordToken(ctx), ctx)
+
+    let isForum = false
+    const isThread = isDiscordThread(channel)
+
+    if (isThread || isForum) {
+      if (!firstThreadMessage) {
+        parent = channel.id
+      }
+
+      if (channel.parent_id) {
+        const parentChannelObj = await getChannel(channel.parent_id, getDiscordToken(ctx), ctx)
+        parentChannel = parentChannelObj.name
+        isForum = isDiscordForum(parentChannelObj)
+      }
+    }
+    // record.parentId means that it's a reply
+    else if (record.message_reference && record.message_reference.message_id) {
+      parent = record.message_reference.message_id
+    }
+
+    await ctx.publishData<IDiscordAPIData>({
+      type: 'channel',
```

Comment:
NIT: let's introduce some enums and use those when publishing data types (channel, member etc)